### PR TITLE
Seperate 'users' table

### PR DESCRIPTION
**Pull request by MichaelSefranek**

### CODE REVIEW

**Comment by MichaelSefranek** [comment]
To be able to separate and reduce repeated information from the names in the Database, and put it into a table.  Our current CSV contains information like this : 

``` 
time,interval,name,stressLevel,submission
1488899559,first_of_day,Mike,3,nothin at all 
```

and we would like to be able to take a name like "Mike", and insert it into students table, so that the name Mike does not have to be repeated for each entry in the Database, rather it will refer to the table of students

**Comment by MichaelSefranek** [comment]
1. [x] Create a new table that contains the names and their keys
2. [x] For every unique name in the Database, replace that name with a unique key
3. [ ] When we try to retrieve a student's information from the Database, make the program refer to the table of students for the name instead of a hard-coded name.
4. [x] If we add a submission to the Database, make sure the name is added to the names table if it doesn't already exist.

**Comment by MichaelSefranek** [comment]
created table in the setup_database.rb file for students.  Next task is to add items to this table.

**Comment by MichaelSefranek** [comment]
Thoughts on the next step: I am looking towards the Submission class, where we sent information to our Database previously.  I want to intercept where we gathered name here, and added it to our database.
```
  def create(params)
    student = Student.new(params[:name])
    new_submission = "'#{Time.now.strftime("%x")}','#{Time.now.to_i}','#{student.last_submission_at}','#{params[:name]}','#{params[:stressLevel]}','#{params[:submission]}'"
    $database.add(new_submission)
  end
```
So here we have `'#{params[:name]}'` being sent to our Submissions table via the Database class. The `add` needs to send  `'#{params[:name]}'` to our `students` table. Then it will send the corresponding `id` for that student to our Submissions table.

Next up I can see that  another additional step is to make sure the name given to the names table is unique compared to all the values in our names table.

**Comment by MichaelSefranek** [comment]
- [x] add name of student to the student table
- [ ] check to see if students name is unique in the table

**Comment by MalastrumDominiSui** [comment]
` PG::UndefinedColumn - ERROR:  column "mike" does not exist` is still the error I am getting.  I tried
```
@conn.exec("INSERT INTO students(name) VALUES(#{student_name})")
```
initially, then I tried to give it quotes, as recommended (unless i misunderstood)
```
    student_name = "\"" + @name + "\""
    @conn.exec("INSERT INTO students(name) VALUES(#{student_name})")
```
which throws the same error for both 
```PG::UndefinedColumn - ERROR:  column "mike" does not exist```

I've tried other forms as well, and it still doesn't makes sense because previously we used 
```
@conn.exec("INSERT INTO submissions (date, time, interval, name, stressLevel, submission) VALUES (#{row})")
``` 
and it worked fine, (where row was a string)

**Comment by MichaelSefranek** [comment]
now `"INSERT INTO students (name) VALUES('#{@name}')"` yields
no implicit conversion of nil into String:

**Comment by MichaelSefranek** [comment]
Figured it out! Thanks to Sumeet for pointing out the obvious.  Refreshing Postico revealed the function actually works.

**Comment by MichaelSefranek** [comment]
next step is going to check to see if the name going in is already inside the database.  My thoughts here is that using our `addname` function within the Student class;
- query the database
- return all names
- only if the new name is not in the current list of names do we add it to our table

**Comment by MichaelSefranek** [comment]
Working on getting a function to get all student names from our student names database.  This is the meat of what I'm doing.  In the database class we already have something useful.

```
  # Get all rows based on a requested header value (EX: header => "names" only returns all names)
  #
  # Removes duplicate entries.
  #
  # Returns an Array of Strings.
  def get_items_by_header(header)
    list = []
    all_items = @conn.exec("SELECT #{header} FROM submissions")
    all_items.each do |row|
      list << row.values
    end
    return list.uniq
  end
```

but the problem to return students from a students table is that this function is specific to submissions.
I am not replacing it, but making a new general function to accept a table name, instead.  I will look at refactoring later.

**Comment by MichaelSefranek** [comment]
changed my mind, before testing I should move all database functionality out of the student class, since hardcoding the connection to the database as an instance variable seems to significantly affect how the student class may operate

**Comment by MichaelSefranek** [comment]
i am now extracting new names, only when they are new as we create submissions.  Next I need to replace or cross-reference the names within the submissions database.

**Comment by MichaelSefranek** [comment]
Writing a test for `name_is_new`, just fixed Gemfile that was missing Rspec.

**Comment by MichaelSefranek** [comment]
Need to run the code to create testing tables for the testing database.

**Comment by MichaelSefranek** [comment]
Got test database running, now working on a teardown.  To make a teardown I need to be able to remove a row from a table by name.  This is probably useful to have as a function in general, not just for testing.  Remove row by value would be useful.

**Comment by MichaelSefranek** [comment]
The tests I wrote for student_spec was a continuation of some of my work with Sumeet this afternoon.  I had not seen tests done in multiple `Rspec.describe` statements, and I was unsure the significance of bundling them like this.  They all pass.  There may be problems if for some reason any student of a these exact names are in, but I am assuming the database is empty to being with.

**Comment by MichaelSefranek** [comment]
So the last piece of this is to have the names column in the Submissions table cross reference in SQL to the students - name column in the students table.

**Comment by MichaelSefranek** [comment]
So more specifically, the students table is composed of a Primary Key, and a name. When we add a name to names table(which is already being done),add its foreign key to the submissions table.  Then we also have to make sure that the whenever name is retrieved from the Submissions database, it will actually return what we want, the actual name and not the foreign key.

File: models/student.rb [hunk 1]
```diff
@@ -1,15 +1,33 @@
-# Contains all functionality relating to student tracking.
+# Contains all functionality relating to interacting with 'students' table.
```
Comment (bdfairbanks) [inline models/student.rb:1]: NOT TRUE!!!!!!!! (yt)

File: services/database.rb [hunk 2]
```diff
@@ -40,6 +40,23 @@ def all_by(table, key, value)
   # Removes duplicate entries.
   #
   # Returns an Array of Strings.
+  def table_items_by_header(header, table)
```
Comment (bdfairbanks) [inline services/database.rb]: NOT NEEDED!!!!!!!!!

Comment (bdfairbanks) [reply services/database.rb]: NOT PUSHED RECENTLY!!!

**Review by bdfairbanks — APPROVED** [review]
LOOKS GOOD!!!!

**Pull request merged**